core/templates/dev/head/pages/admin/roles_tab/RoleGraphDirective.js

```diff
@@ -0,0 +1,124 @@
+// Copyright 2017 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Directive for displaying Role graph.
+ */
+
+oppia.directive('roleGraph', [
+  'UrlInterpolationService', function(UrlInterpolationService) {
+    return {
+      restrict: 'E',
+      scope: {
+        // An object with these keys:
+        //  - 'nodes': An object whose keys are node ids and whose values are
+        //             node labels
+        //  - 'links': A list of objects with keys:
+        //            'source': id of source node
+        //            'target': id of target node
+        //            'linkProperty': property of link which determines how
+        //              it is styled (styles in linkPropertyMapping). If
+        //              linkProperty or corresponding linkPropertyMatching
+        //              is undefined, link style defaults to the gray arrow.
+        //  - 'initStateId': The initial state id
+        //  - 'finalStateIds': The list of ids corresponding to terminal states
+        //             (i.e., those whose interactions are terminal).
+        graphData: '=',
+        // A boolean value to signify whether graphData is completely loaded.
+        graphDataLoaded: '@'
+      },
+      templateUrl: UrlInterpolationService.getDirectiveTemplateUrl(
+        '/pages/admin/roles_tab/role_graph_directive.html'),
+      controller: [
+        '$scope', '$element', '$timeout', '$filter', 'StateGraphLayoutService',
+        'MAX_NODES_PER_ROW', 'MAX_NODE_LABEL_LENGTH',
+        function(
+            $scope, $element, $timeout, $filter, StateGraphLayoutService,
+            MAX_NODES_PER_ROW, MAX_NODE_LABEL_LENGTH) {
+          // A rough upper bound for the width of a single letter, in pixels,
+          // to use as a scaling factor to determine the width of graph nodes.
+          // This is not an entirely accurate description because it also takes
+          // into account the horizontal whitespace between graph nodes.
+          var LETTER_WIDTH_IN_PIXELS = 10.5;
```

Comment:
could these stuff also be factored out into a function that gives the graph width and height?